### PR TITLE
Fix argmax/min to return first index

### PR DESCRIPTION
**Pull request by choidongyeon**
Addresses the issue mentioned in https://github.com/pytorch/pytorch/issues/10370 and https://github.com/pytorch/pytorch/issues/37921, where PyTorch currently returns the last index value rather than the first index value for argmax/min.

Example output:

```
t = torch.ones(3, 4).cpu()
t[1, 0] = 0
print("argmax :")
print(t.argmax(1, keepdim=True))
print("argmin :")
print(t.argmin(1, keepdim=True))
```
Output:
```
argmax :
tensor([[0],
        [1],
        [0]])
argmin :
tensor([[0],
        [0],
        [0]])
```

Example 2:

```
t = torch.eye(3)
print("argmax :")
print(t.argmax(1, keepdim=True))
print("argmin :")
print(t.argmin(1, keepdim=True))
```
Output:
```
argmax :
tensor([[0],
        [1],
        [2]])

argmin :
tensor([[1],
        [0],
        [0]])
```

### CODE REVIEW

**Review by mruberry — CHANGES_REQUESTED** [review]
Hey @choidongyeon, thanks for the submission! 

Can you comment on why this change produces the desired result, and can you add a couple tests verifying that the behavior matches NumPy's? For example, testing a variety of different inputs with tensors of different sizes, dtypes, and values.

**Comment by choidongyeon** [comment]
> Can you comment on why this change produces the desired result

Hey @mruberry, thanks for taking a look! 

I think the key is understanding what is going on here in [binary_kernel_reduce](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/Reduce.h#L213) 

```
acc = ops.reduce(acc, *(data_t*)in, begin + i);
```

If you follow the code starting with [argmax_kernel_impl](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L267), you can see that from there, binary_kernel_reduce calls on [ArgMaxOps](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SharedReduceOps.h#L369), which uses [the reduction operation happening in this line](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SharedReduceOps.h#L346):

```
comp_t{}(arg.first, val) ? arg : arg_t(val, idx)
```

Here, each pair in ```arg``` is of a (current max valued element, the element's corresponding index). The comparison happens in [GreaterOrNan](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/SharedReduceOps.h#L331), which was originally written as
```
return at::_isnan(a) || a > b;
```
This would mean that ```GreaterOrNan``` will only return true if `a` (or ```arg.first```) is strictly greater than `b` (or ```val```). If the two values were of equal value, it would return false. Going back to the above line in ```binary_kernel_reduce```, we can see that ```acc``` would then always point to the latest (max value, corresponding index) pair.

The change I'm proposing in this PR is to allow ```GreaterOrNan``` to return true when ```arg.first``` and ```val``` are equal:
```
return at::_isnan(a) || a >= b;
```
That way, the (value, index) pair in ```binary_kernel_reduce``` (stored as ```acc```) remains unchanged unless the new encountered value is strictly greater than the current stored value.

Similar reasoning can be applied for argmin. Does this make sense to you?

> Can you add a couple tests verifying that the behavior matches NumPy's? For example, testing a variety of different inputs with tensors of different sizes, dtypes, and values.

Sure! Is there anywhere in particular that I should add these tests? It seems like [test/test_torch.py](https://github.com/pytorch/pytorch/blob/master/test/test_torch.py) might be the appropriate place but wanted to check with you first.

**Comment by mruberry** [comment]
Reasoning makes sense ;) (although the function names should be updated to reflect they now return true if the value is nan or greater/lesser than or equal to).

test_torch.py is the correct place for tests. There may be existing argax/argmin tests you can update, of you may want to write a new one for this behavior specifically. Be sure to test enough varieties that we are confident this change addresses the issue regardless of the arguments. I'm concerned that it seems very dependent on the structure of our CPU and GPU reductions, and those structures are not guaranteed to be the same.

**Comment by choidongyeon** [comment]
> (although the function names should be updated to reflect they now return true if the value is nan or greater/lesser than or equal to).

How about ```GreaterEqualOrNan``` and ```LessEqualOrNan```?



> I'm concerned that it seems very dependent on the structure of our CPU and GPU reductions, and those structures are not guaranteed to be the same.

Got it, will try to come up with as many cases as I can and will update the PR!

**Comment by bonlime** [comment]
I've just found the same undesired behavior. Check #37921 
@choidongyeon could you also make sure that `cuda` argmax would return first index?

**Comment by choidongyeon** [comment]
> I've just found the same undesired behavior. Check #37921
> @choidongyeon could you also make sure that `cuda` argmax would return first index?

Hey @bonlime, I just read the issue you created. It suggests that `cuda` argmax is fine and it's cpu argmax that yields the issue? In which case, I just checked. With my changes :
```
In [1]: import torch                                                                                                                                  

In [2]: torch.max(torch.ones(1, 4), dim=1)                                                                                                                
Out[2]: 
torch.return_types.max(
values=tensor([1.]),
indices=tensor([0]))
```

**Comment by bonlime** [comment]
@choidongyeon you are right. It looks like after your change CPU and CUDA version would match. Good job.

**Comment by mruberry** [comment]
Hey @choidongyeon! So I had a chance to talk to a colleague about this, and we have two concerns:

- as mentioned, this PR depends heavily on the structure of reductions, which is not guaranteed to be the same across platforms or hardware, so even if the approach will work it'll be extremely challenging to write tests for (the current test failures on some platforms may be an indication of this) 
- we are concerned about users relying on this behavior because we would like the ability to prioritize performance over it (at least for now), and so may have to change the structure of kernels again the future

I appreciate your working through a simple fix that will work in a lot of cases, but this issue is probably too thorny to resolve.

**Comment by choidongyeon** [comment]
Hi @mruberry! Yeah, I was looking into the test failures, and it seems that with the addition of tests, I have been able to show that my solution would not work for nd arrays. For the 3d array/tensor
```
[[[-12.   8. -14.]
  [ 12.  12.  12.]
  [  8.   9.   9.]]

 [[ -8.  -5.  -7.]
  [  4.   6.   2.]
  [ 15. -19. -13.]]]
```
my PR solution yields (for argmax axis 2):
```
pytorch argmax: tensor([[1, 0, 2],
        [1, 1, 0]], device='cuda:0')
numpy argmax :  tensor([[1, 0, 1],
        [1, 1, 0]], device='cuda:0')
```
It seems that something still isn't being addressed, which as you suggested, is likely due to the fact that this solution relies on the current structure of reductions.

> * we are concerned about users relying on this behavior because we would like the ability to prioritize performance over it (at least for now), and so may have to change the structure of kernels again the future

I definitely agree with you that performance should be priority over trying to have outcomes consistent with numpy's conventions.  But I'm wondering if we wanted to at least change to be a little more consistent with numpy (at least for now) since the [inconsistency is already documented here](https://pytorch.org/docs/master/generated/torch.max.html#torch.max) and this wouldn't impact performance itself?

Let me know what you think and what my next steps should be (i.e. should I close the PR?). 😄

**Comment by mruberry** [comment]
> Hi @mruberry! Yeah, I was looking into the test failures, and it seems that with the addition of tests, I have been able to show that my solution would not work for nd arrays. For the 3d array/tensor
> 
> ```
> [[[-12.   8. -14.]
>   [ 12.  12.  12.]
>   [  8.   9.   9.]]
> 
>  [[ -8.  -5.  -7.]
>   [  4.   6.   2.]
>   [ 15. -19. -13.]]]
> ```
> 
> my PR solution yields (for argmax axis 2):
> 
> ```
> pytorch argmax: tensor([[1, 0, 2],
>         [1, 1, 0]], device='cuda:0')
> numpy argmax :  tensor([[1, 0, 1],
>         [1, 1, 0]], device='cuda:0')
> ```
> 
> It seems that something still isn't being addressed, which as you suggested, is likely due to the fact that this solution relies on the current structure of reductions.
> 
> > * we are concerned about users relying on this behavior because we would like the ability to prioritize performance over it (at least for now), and so may have to change the structure of kernels again the future
> 
> I definitely agree with you that performance should be priority over trying to have outcomes consistent with numpy's conventions. But I'm wondering if we wanted to at least change to be a little more consistent with numpy (at least for now) since the [inconsistency is already documented here](https://pytorch.org/docs/master/generated/torch.max.html#torch.max) and this wouldn't impact performance itself?
> 
> Let me know what you think and what my next steps should be (i.e. should I close the PR?). 😄

It was a really cool idea, and I'm glad we got to look at the tests ;)

Yep, I think we just close this one. Looking forward to the next PR, though!

**Comment by choidongyeon** [comment]
Awesome, closing it now. Thanks for taking a look at it so quickly!
